From c89195855d3dec5210b1d50f0fe28b9bce410b40 Mon Sep 17 00:00:00 2001
From: svenkatachalam <svenkatachalam@maxlinear.com>
Date: Tue, 23 Jul 2024 21:08:24 +0800
Subject: [PATCH] WLANRTSYS-82491 GCMP-256 is not enabled by default in
 Hostapd-ng

As per 11be spec, there should be mandatory support for GCMP-256 cipher suite.

Fix:
Enabling GCMP256 cipher suite.

FIXUP to WLANRTSYS-60843 MLO changes: AP MLD configuration.
---
 hostapd/mxl_config.c | 12 ++++++++++++
 src/ap/ap_config.c   |  5 +++++
 src/ap/mxl_config.h  |  1 +
 3 files changed, 18 insertions(+)

diff --git a/hostapd/mxl_config.c b/hostapd/mxl_config.c
index ebc4b177e..823537b0e 100644
--- a/hostapd/mxl_config.c
+++ b/hostapd/mxl_config.c
@@ -2060,3 +2060,15 @@ int mxl_hostapd_config_check(struct hostapd_config *conf, int full_config)
 	return 0;
 }
 
+int mxl_hostapd_config_check_bss(struct hostapd_bss_config *bss, struct hostapd_config *conf, int full_config)
+{
+#ifdef CONFIG_IEEE80211BE
+	if (full_config && conf->ieee80211be && !bss->disable_11be &&
+		((bss->wpa_key_mgmt & WPA_KEY_MGMT_OWE) || wpa_key_mgmt_sae(bss->wpa_key_mgmt))) {
+			bss->rsn_pairwise |= WPA_CIPHER_GCMP_256;
+			wpa_printf(MSG_WARNING,
+				"GCMP-256 cipher suite is force-enabled due to 802.11be is enabled");
+	}
+#endif /* CONFIG_IEEE80211BE */
+	return 0;
+}
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index cf6612b3d..e0101374e 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -1635,6 +1635,11 @@ static int hostapd_config_check_bss(struct hostapd_bss_config *bss,
 		return -1;
 	}
 
+#ifdef CONFIG_VENDOR_MXL
+	if (mxl_hostapd_config_check_bss(bss, conf, full_config))
+		return -1;
+#endif /* CONFIG_VENDOR_MXL */
+
 	return 0;
 }
 
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index 5c3694d03..882ea9c2b 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -418,6 +418,7 @@ void mxl_hostapd_config_free_bss(struct hostapd_bss_config *bss);
 int mxl_hostapd_config_check(struct hostapd_config *conf, int full_config);
 void mxl_hostapd_config_free(struct hostapd_config *conf);
 int mxl_hostapd_config_check_bss_6g(struct hostapd_bss_config *bss);
+int mxl_hostapd_config_check_bss(struct hostapd_bss_config *bss, struct hostapd_config *conf, int full_config);
 
 int mxl_hostapd_parse_intlist(int **int_list, char *val);
 
-- 
2.43.0

